tunnistamo/tests/test_auth_error_redirect.py

```diff
@@ -0,0 +1,177 @@
+    ('', 'interaction_required', _('Authentication failed')),
+    ('unknown error code', 'interaction_required', _('Authentication failed')),
+    ('interaction_required', 'interaction_required', _('Authentication failed')),
+    ('access_denied', 'access_denied', _('Authentication cancelled or failed')),
+])
+def test_redirect_to_client_after_social_auth_error(
+    django_user_model,
+    oidcclient_factory,
+    original_error,
+    expected_error,
+    expected_error_description
+):
+    django_client = CancelSocialAuthBackendRedirectClient()
+
+    # Start the OIDC flow.
+    state = get_random_string()
+    oidc_client = _start_oidc_authorize(django_client, oidcclient_factory, state=state)
+
+    # Return the user from the social auth with an error.
+    response = _request_social_auth_complete_with_error(
+        django_client,
+        backend_name=HelsinkiTunnistus.name,
+        error=original_error,
+    )
+
+    # Verify that the oidc provider would redirect the user back to the client
+    expected_redirect_url = '{}?{}'.format(oidc_client.redirect_uris[0], urlencode({
+        'error': expected_error,
+        'error_description': expected_error_description,
+        'state': state,
+    }))
+    assert response.status_code == 302
+    assert response.url == expected_redirect_url
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('on_error_redirect', (True, False))
+def test_on_auth_error_redirect_to_client_setting(
+    settings,
+    django_user_model,
+    oidcclient_factory,
+    on_error_redirect,
+):
+    settings.AUTHENTICATION_BACKENDS = settings.AUTHENTICATION_BACKENDS + (
+        'tunnistamo.tests.conftest.DummyFixedOidcBackend',
+    )
+    settings.SOCIAL_AUTH_DUMMYFIXEDOIDCBACKEND_ON_AUTH_ERROR_REDIRECT_TO_CLIENT = on_error_redirect
+    reload_social_django_utils()
+
+    django_client = CancelSocialAuthBackendRedirectClient()
+
+    # Start the OIDC flow.
+    oidc_client = _start_oidc_authorize(django_client, oidcclient_factory)
+
+    # Return the user from the social auth with an error.
+    response = _request_social_auth_complete_with_error(django_client)
+
+    assert response.status_code == 302
+
+    if on_error_redirect:
+        # Verify that the oidc provider would redirect the user back to the client
+        expected_redirect_url = '{}?{}'.format(oidc_client.redirect_uris[0], urlencode({
+            'error': 'interaction_required',
+            'error_description': _('Authentication failed'),
+        }))
+        assert response.url == expected_redirect_url
+    else:
+        # Without the on error auth redirect setting the user should be redirected to
+        # the login view
+        assert response.url.startswith('/login/')
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('on_error_redirect', (True, False))
+def test_should_not_redirect_to_oidc_client_if_the_next_parameter_is_not_to_an_oidc_endpoint(
+    settings,
+    django_user_model,
+    oidcclient_factory,
```

Comment:
`django_user_model` and `oidcclient_factory` are not used here. Can they be removed?